src/Modules/Visualization/ShowField.cc

```diff
@@ -448,754 +538,293 @@ void GeometryBuilder::renderFacesLinear(
   int numAttributes = 3; //intially 3 because we will atleast be rendering verticies (vec3's)
 
   mesh->size(numFaces);
-  if (numFaces == 0)
-    return;
+  if (numFaces == 0) return;
 
-  bool withNormals = state.get(RenderState::USE_NORMALS);
-  if (withNormals)
+  bool useNormals = state.get(RenderState::USE_NORMALS);
+  bool useFaceNormals = state.get(RenderState::USE_FACE_NORMALS) && mesh->has_normals();
+  bool invertNormals = state_->getValue(ShowField::FaceInvertNormals).toBool();
+  if (useNormals)
   {
     numAttributes += 3;
     mesh->synchronize(Mesh::NORMALS_E);
   }
 
-  bool invertNormals = state_->getValue(ShowField::FaceInvertNormals).toBool();
+  bool useColorMap = (fld->basis_order() >= 0 && state.get(RenderState::USE_COLORMAP));
   ColorScheme colorScheme = ColorScheme::COLOR_UNIFORM;
-  std::vector<double> svals;
-  std::vector<Vector> vvals;
-  std::vector<Tensor> tvals;
-  std::vector<ColorRGB> face_colors;
 
-  if (fld->basis_order() < 0 || state.get(RenderState::USE_DEFAULT_COLOR))
-  {
-    colorScheme = ColorScheme::COLOR_UNIFORM;
-  }
-  else if (state.get(RenderState::USE_COLORMAP))
+  auto realColorMap = colorMap.get();
+  if(!realColorMap) realColorMap = StandardColorMapFactory::create();
+
+  auto map = StandardColorMapFactory::create(realColorMap->getColorMapName(),
+    realColorMap->getColorMapResolution(), realColorMap->getColorMapShift(),
+    realColorMap->getColorMapInvert());
+
+  auto coordinateMap = StandardColorMapFactory::create("Grayscale", 256, 0, false,
+    realColorMap->getColorMapRescaleScale(), realColorMap->getColorMapRescaleShift());
+
+  if (useColorMap)
   {
-    numAttributes += 4;
+    numAttributes += 2;
     colorScheme = ColorScheme::COLOR_MAP;
   }
-  else // if (fld->basis_order() >= 0)
-  {
-    numAttributes += 4;
-    colorScheme = ColorScheme::COLOR_IN_SITU;
-  }
 
   Point idpt;
   VMesh::Face::iterator fiter, fiterEnd;
   VMesh::Node::array_type nodes;
-  mesh->begin(fiter);
-  mesh->end(fiterEnd);
   mesh->get_nodes(nodes, *fiter);
   mesh->get_point(idpt, nodes[0]);
+  mesh->begin(fiter);
+  mesh->end(fiterEnd);
 
-  // Three 32 bit ints for each triangle to index into the VBO (triangles = verticies - 2)
-  size_t iboSize = static_cast<size_t>(mesh->num_faces() * sizeof(uint32_t) * (nodes.size() - 2) * 3);
-  size_t vboSize = static_cast<size_t>(mesh->num_faces() * sizeof(float) * nodes.size() * numAttributes);
-
-  // Construct VBO and IBO that will be used to render the faces. Once again, IBOs are not strictly
-  // needed. But, we may be able to optimize this code somewhat.
-  // todo Switch to unique_ptrs and move semantics.
-  std::shared_ptr<spire::VarBuffer> iboBufferSPtr(new spire::VarBuffer(iboSize));
-  std::shared_ptr<spire::VarBuffer> vboBufferSPtr(new spire::VarBuffer(vboSize));
+  int numNodes = nodes.size();
+  bool useQuads = (numNodes == 4);
+  int writeCase = ((int)useQuads << 2) | ((int)useNormals << 1) | ((int)useColorMap);
```

Comment:
This should definitely be a function returning a typed enum value with the correct label that makes the switch readable

Reply:
Won't that kinda have to look like this?

  inline int getWriteCase(bool writeQuads, bool writeNormals, bool writeTexCoords)
  {
    if(writeQuads)
    {
      if(writeNormals)
      {
        if(writeTexCoords)
        {
          return QUAD_NORMALS_TEXCOORDS;
        }
        else
        {
          return QUAD_NORMALS;
        }
      }
      else
      {
        if(writeTexCoords)
        {
          return QUAD_TEXCOORDS;
        }
        else
        {
          return QUAD;
        }
      }
    }
    else
    {
      if(writeNormals)
      {
        if(writeTexCoords)
        {
          return TRI_NORMALS_TEXCOORDS;
        }
        else
        {
          return TRI_NORMALS;
        }
      }
      else
      {
        if(writeTexCoords)
        {
          return TRI_TEXCOORDS;
        }
        else
        {
          return TRI;
        }
      }
    }
  }

Reply:
No, you can set up the enum class to have the same values that the `int` has now, then just construct it:
```
EnumTypeName writeCase(bool writeQuads, bool writeNormals, bool writeTexCoords)
{
  return EnumTypeName(((int)useQuads << 2) | ((int)useNormals << 1) | ((int)useColorMap));
}
```

Reply:
If it had to look like that do you think I'd suggest it? 😝

Reply:
Can we talk about this one in person?

Reply:
That way the binary values are hidden and turn into named constants (remember no magic numbers!). This is done all the time in Qt with "flag enums" that have values of powers of 2, and are bit-shifted and bit-arithmetic'ed together in different ways

Reply:
https://stackoverflow.com/questions/1448396/how-to-use-enums-as-flags-in-c